src/ngx_http_lua_semaphore.c

```diff
@@ -0,0 +1,433 @@
+
+    sem = coctx->data;
+
+    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, sem->log, 0,
+                   "ngx_http_lua_semaphore_cleanup");
+
+    ngx_del_timer(&coctx->sleep);
+
+    q = &coctx->sem_wait_queue;
+
+    ngx_queue_remove(q);
+    sem->wait_count--;
+    coctx->cleanup = NULL;
+}
+
+
+void
+ngx_http_lua_ffi_semaphore_handler(ngx_event_t *ev)
+{
+    ngx_http_lua_semaphore_t          *sem;
+    ngx_http_request_t                *r;
+    ngx_http_lua_ctx_t                *ctx;
+    ngx_http_lua_co_ctx_t             *wait_co_ctx;
+    ngx_connection_t                  *c;
+    ngx_queue_t                       *q;
+
+    sem = ev->data;
+
+    while (!ngx_queue_empty(&sem->wait_queue) && sem->count > 0) {
+
+         q = ngx_queue_head(&sem->wait_queue);
+         ngx_queue_remove(q);
+         wait_co_ctx = ngx_queue_data(q, ngx_http_lua_co_ctx_t, sem_wait_queue);
+
+         sem->count--;
+         sem->wait_count --;
+         ngx_del_timer(&wait_co_ctx->sleep);
+         wait_co_ctx->cleanup = NULL;
+
+         r = ngx_http_lua_get_req(wait_co_ctx->co);
+
+         ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+         if (ctx == NULL) {
+             continue;
+         }
+
+         wait_co_ctx->cleanup = NULL;
+         ctx->cur_co_ctx = wait_co_ctx;
+         wait_co_ctx->sem_resume_status = NGX_HTTP_LUA_SEM_WAIT_SUCC;
+         c = r->connection;
+
+         if (ctx->entered_content_phase) {
+             (void) ngx_http_lua_semaphore_resume(r);
+
+         } else {
+             ctx->resume_handler = ngx_http_lua_semaphore_resume;
+             ngx_http_core_run_phases(r);
+         }
+
+         ngx_http_run_posted_requests(c);
+
+    }
+
+    /*after dealing semaphore post event */
+    sem->in_post_event = 0;
+
+}
+
+
+static void
+ngx_http_lua_sem_timeout_handler(ngx_event_t *ev)
+{
+    ngx_http_lua_co_ctx_t       *wait_co_ctx;
+    ngx_http_request_t          *r;
+    ngx_http_lua_ctx_t          *ctx;
+    ngx_connection_t            *c;
+    ngx_http_lua_semaphore_t    *sem;
+
+    wait_co_ctx = ev->data;
+
+    r = ngx_http_lua_get_req(wait_co_ctx->co);
+
+    dd("ngx_http_lua_sem_timeout_handler timeout coctx:%p", wait_co_ctx);
+
+    sem = wait_co_ctx->data;
+    sem->wait_count --;
+    wait_co_ctx->cleanup = NULL;
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+    if (ctx == NULL) {
+        return;
+    }
+
+    ctx->cur_co_ctx = wait_co_ctx;
+
+    wait_co_ctx->sem_resume_status = NGX_HTTP_LUA_SEM_WAIT_TIMEOUT;
+
+    c = r->connection;
+
+    if (ctx->entered_content_phase) {
+        (void) ngx_http_lua_semaphore_resume(r);
+
+    } else {
+         ctx->resume_handler = ngx_http_lua_semaphore_resume;
+         ngx_http_core_run_phases(r);
+    }
+
+    ngx_http_run_posted_requests(c);
+}
+
+
+void
+ngx_http_lua_ffi_sem_gc(ngx_http_lua_semaphore_t *sem)
+{
+    ngx_log_debug1(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, 0,
+     "ngx_http_lua_ffi_sem_gc %p", sem);
```

Comment:
Style: continued argument list should be aligned up with the position right after the `(` char in the previous line.